components/squeezelite/tas57xx/dac_57xx.c

```diff
@@ -130,9 +139,68 @@ static void deinit(void)	{
 
 /****************************************************************************************
  * change volume
+ * Parameter going from 0 to 65536
+ * TAS allows from 254 = -103 dB (negative maximum) to 0 = +24 dB (positive maximum)
+ * TAS default is 0 dB = 0x30 = 48 no attenuation
+ * TAS control from -103 dB to 0 dB
+ * A Value of 255 means negative infinite (Mute)
+ * TAS gain will be implemented from
+ * 
+ * -infinity (Mute) = 0xff = 255
+ * -74.5 dB = 0xd3 = 197
+ *   .
+ *   .
+ *   .
+ *  7 dB = 0x42 = 34
+ * 
+ * representiong from 0 to 65536 the input
+ * 
+ * This scaling might be changed. We start at -74.5 dB, lower values result
+ * in nearly not hearable output at 15% Volume. This scaling is adopted to a
+ * output.gainL/r of FIXED_ONE/2 This enalbes the eqaulizer some range for adjustments
+ * berfore clipping too early. We will use 1 bit of resolution.
+ * 
+ * representiong from 0 to 65536 the input
  */
+
 static bool volume(unsigned left, unsigned right) { 
-	return false; 
+	esp_err_t ret = ESP_OK;
+	i2c_cmd_handle_t i2c_cmd = i2c_cmd_link_create();
+
+	ESP_LOGI(TAG, "Volume is requested to: L: %d, R: %d", left, right);
+
+	if (left > 0) {
+		left = (unsigned) min(round(- log10(((double) left) / (double) FIXED_ONE) * 45) + 34, 255);
+	} else {
+		left = 0xff;
+	}
+	i2c_master_start(i2c_cmd);
+	i2c_master_write_byte(i2c_cmd, tas57_addr | I2C_MASTER_WRITE, I2C_MASTER_NACK);
+	i2c_master_write_byte(i2c_cmd, 0x3d, I2C_MASTER_NACK);
+	i2c_master_write_byte(i2c_cmd, (uint8_t) left, I2C_MASTER_NACK);
+	ESP_LOGI(TAG, "i2c write %x at %u", 0x3d, (uint8_t) left);
+
+	if (left > 0) {
```

Comment:
Are you sure about that?

Reply:
Absolutely not! Very good eyes and review. It was a copy / paste error. Removed.